dali/pipeline/operators/arithmetic/arithmetic_test.cc

```diff
@@ -0,0 +1,338 @@
+
+namespace dali {
+
+TEST(ArithmeticOps, ExpressionTree) {
+  std::string expr = "mul(&0 &1)";
+  auto result = ParseExpressionString(expr);
+  auto &result_ref = *result;
+  ASSERT_EQ(result_ref.GetNodeType(), NodeType::Function);
+  ASSERT_EQ(result_ref.GetSubexpressionCount(), 2);
+  ASSERT_EQ(result_ref.GetOp(), "mul");
+  ASSERT_EQ(result_ref[0].GetNodeType(), NodeType::Tensor);
+  ASSERT_EQ(result_ref[1].GetNodeType(), NodeType::Tensor);
+  ASSERT_EQ(dynamic_cast<ExprTensor &>(result_ref[0]).GetMappedInput(), 0);
+  ASSERT_EQ(dynamic_cast<ExprTensor &>(result_ref[1]).GetMappedInput(), 1);
+}
+
+TEST(ArithmeticOps, ExpressionTreeComplex) {
+  std::string expr = "div(sub(&42 &2) &1)";
+  auto result = ParseExpressionString(expr);
+  auto &result_ref = *result;
+  ASSERT_EQ(result_ref.GetNodeType(), NodeType::Function);
+  ASSERT_EQ(result_ref.GetSubexpressionCount(), 2);
+  EXPECT_EQ(result_ref.GetOp(), "div");
+  ASSERT_EQ(result_ref[0].GetNodeType(), NodeType::Function);
+  ASSERT_EQ(result_ref[0].GetSubexpressionCount(), 2);
+  EXPECT_EQ(result_ref[0].GetOp(), "sub");
+  ASSERT_EQ(result_ref[0][0].GetNodeType(), NodeType::Tensor);
+  EXPECT_EQ(dynamic_cast<ExprTensor &>(result_ref[0][0]).GetMappedInput(), 42);
+  ASSERT_EQ(result_ref[0][1].GetNodeType(), NodeType::Tensor);
+  EXPECT_EQ(dynamic_cast<ExprTensor &>(result_ref[0][1]).GetMappedInput(), 2);
+  ASSERT_EQ(result_ref[1].GetNodeType(), NodeType::Tensor);
+  EXPECT_EQ(dynamic_cast<ExprTensor &>(result_ref[1]).GetMappedInput(), 1);
+}
+
+TEST(ArithmeticOps, TreePropagation) {
+  std::string expr_str = "div(sub(&0 &1) $2:int32)";
+  auto expr = ParseExpressionString(expr_str);
+  auto &expr_ref = *expr;
+  HostWorkspace ws;
+  std::shared_ptr<TensorVector<CPUBackend>> in[3];
+  for (auto &ptr : in) {
+    ptr = std::make_shared<TensorVector<CPUBackend>>();
+    ptr->Resize({{1}, {2}});
+  }
+  in[0]->set_type(TypeInfo::Create<uint8_t>());
+  in[1]->set_type(TypeInfo::Create<int16_t>());
+  in[2]->set_type(TypeInfo::Create<int32_t>());
+  ws.AddInput(in[0]);
+  ws.AddInput(in[1]);
+  ws.AddInput(in[2]);
+
+  auto result_type = PropagateTypes<CPUBackend>(expr_ref, ws);
+  auto result_shape = PropagateShapes<CPUBackend>(expr_ref, ws);
+  auto expected_shpe = kernels::TensorListShape<>{{1}, {2}};
+  EXPECT_EQ(result_type, DALIDataType::DALI_INT32);
+  EXPECT_EQ(result_shape, expected_shpe);
+  EXPECT_EQ(expr_ref.GetNodeDesc(), "div:T:int32(FT:int16 CC:int32)");
+  EXPECT_EQ(expr_ref.GetOutputDesc(), "FT:int32");
+  EXPECT_EQ(expr_ref.GetNodeDesc(), "div:T:int32(FT:int16 CC:int32)");
+  EXPECT_EQ(expr_ref.GetOutputDesc(), "FT:int32");
+  EXPECT_EQ(expr_ref[0].GetNodeDesc(), "sub:T:int16(TT:uint8 TT:int16)");
+  EXPECT_EQ(expr_ref[0].GetOutputDesc(), "FT:int16");
+  EXPECT_EQ(expr_ref[1].GetNodeDesc(), "CC:int32");
+  EXPECT_EQ(expr_ref[1].GetOutputDesc(), "CC:int32");
+}
+
+// namespace {
+
+inline bool operator==(const TileDesc &l, const TileDesc &r) {
```

Comment:
Maybe it should be global, not only test specific function.

Reply:
+1